scripts/machine_translation/train_gnmt.py

```diff
@@ -190,43 +214,94 @@ def evaluate(data_loader):
     return avg_loss, real_translation_out
 
 
+def clip_global_norm(arrays, max_norm, context_len=len(context)):
+    """Enable the clip_global_norm function at multi contexts.
+       Reference to gluon.utils.clip_global_norm()
+
+    """
+    def _norm(array):
+        if array.stype == 'default':
+            x = array.reshape((-1,))
+            return ndarray.dot(x, x)
+        return array.norm().square()
+
+    assert len(arrays) > 0
+    assert context_len >= 1
+
+    ctx = arrays[0].context
+    arrays_at_all_context = arrays[0]
```

Comment:
As you use input `grads = [p.grad(c) for p in model.collect_params().values() for c in context]` in line 290, this is the grad of some parameter at the first context. In line 236, you sum up the gradients of this parameter at all contexts, and the gradient at the first context is doubled.

Reply:
The grads is 1-D list. So you have 'arrays_at_all_context == grads[0] '.